File: server/query-builders.ts
import { DISPLAY_FIELDS, type EnvironmentConfig, type LogKind, type SearchInput } from "./domain.js";

const INDEX_PATTERN = /^[a-zA-Z0-9._,*-]+$/;
const MAX_PAGE_DEPTH = 10_000;

const literal = (value: string): string =>
  value.trim().replaceAll("\\", "\\\\").replaceAll('"', '\\"').replaceAll("\n", " ");

const index = (value: string): string => {
  if (!INDEX_PATTERN.test(value)) {
    throw new Error("索引名称包含不允许的字符");
  }
  return value;
};

const timeRange = (field: string, input: SearchInput): string =>
  `${field} >= "${literal(input.startTime)}" AND ${field} < "${literal(input.endTime)}"`;

const addLike = (conditions: string[], field: string, value?: string): void => {
  if (value?.trim()) conditions.push(`${field} LIKE "*${literal(value)}*"`);
};

const addExactOrLike = (
  conditions: string[],
  field: string,
  value: string | undefined,
  looksComplete: (candidate: string) => boolean
): void => {
  if (!value?.trim()) return;
  const normalized = literal(value);
  conditions.push(looksComplete(value.trim()) ? `${field} == "${normalized}"` : `${field} LIKE "*${normalized}*"`);
};

const fullTransactionId = (value: string): boolean => value.length >= 24 && /\.[puds]_\d+_/i.test(value);
const fullTraceId = (value: string): boolean => /^[a-f\d]{16,64}$/i.test(value);

const resolveIndex = (input: SearchInput, environment: EnvironmentConfig): string => {
  if (input.kind === "transaction") return index(environment.txnlstIndex);
  if (input.kind === "application" || input.kind === "ecp") return index(environment.applogIndex);
  if (!input.index) throw new Error("通用日志必须选择索引");

  const allowed = new Set([
    environment.txnlstIndex,
    environment.txntrcIndex,
    environment.applogIndex,
    environment.apmIndex ?? "traces-apm*"
  ]);
  if (!allowed.has(input.index)) throw new Error("该索引未在当前环境中配置");
  return index(input.index);
};

const buildTransactionConditions = (input: SearchInput): string[] => {
  // UAT writes the ingestion time reliably, but leaves ecp.txn.timestamp empty on
  // some transaction documents. The Java client has always used @timestamp here.
  const conditions = [timeRange("@timestamp", input)];
  addExactOrLike(conditions, "ecp.txn.id", input.txnId, fullTransactionId);
  addExactOrLike(conditions, "ecp.txn.trace", input.traceId, fullTraceId);
  addLike(conditions, "ecp.txn.no", input.txnNo);
  addLike(conditions, "ecp.txn.business", input.business);
  addLike(conditions, "ecp.txn.service", input.service);
  addLike(conditions, "ecp.txn.message.code", input.messageCode);
  addLike(conditions, "ecp.txn.message.info", input.messageInfo);
  addLike(conditions, "ecp.txn.node", input.node);

  if (input.status === "SUCCESS") conditions.push('RIGHT(ecp.txn.message.code, 5) == "00000"');
  if (input.status === "FAIL") conditions.push('RIGHT(ecp.txn.message.code, 5) != "00000"');
  if (input.minDurationMs && input.minDurationMs > 0) {
    conditions.push(`TO_INTEGER(ecp.txn.duration) >= ${Math.floor(input.minDurationMs)}`);
  }
  return conditions;
};

const buildLogConditions = (input: SearchInput): string[] => {
  const timestamp = input.kind === "generic" ? "@timestamp" : "ecp.log.timestamp";
  const conditions = [timeRange(timestamp, input)];
  addLike(conditions, "ecp.log.application", input.application);
  addLike(conditions, "ecp.log.level", input.level);
  if (input.kind === "ecp") addLike(conditions, "ecp.log.file", input.file);

  if (input.keyword?.trim()) {
    const fields = input.kind === "generic"
      ? ["message", "ecp.log.application", "ecp.log.thread", "trace.id"]
      : ["message", "ecp.log.thread", "trace.id", "host.name"];
    const keywordConditions = fields.map((field) => `${field} LIKE "*${literal(input.keyword!)}*"`);
    conditions.push(`(${keywordConditions.join(" OR ")})`);
  }
  return conditions;
};

export const buildSearchQuery = (
  input: SearchInput,
  environment: EnvironmentConfig,
  exportAll = false
): string => {
  const source = resolveIndex(input, environment);
  const conditions = input.kind === "transaction"
    ? buildTransactionConditions(input)
    : buildLogConditions(input);
  const timestamp = input.kind === "generic" ? "@timestamp" : "ecp.log.timestamp";
  const limit = exportAll
    ? 20_000
    : Math.min(input.page * input.pageSize, MAX_PAGE_DEPTH);
  const keep = DISPLAY_FIELDS[input.kind].join(", ");

  if (input.kind === "transaction") {
    // Keep the familiar business-time column populated when legacy UAT records
    // contain only the canonical ingest timestamp.
    return `FROM ${source} | WHERE ${conditions.join(" AND ")} | SORT @timestamp DESC | LIMIT ${limit} | EVAL ecp.txn.timestamp = COALESCE(ecp.txn.timestamp, @timestamp) | KEEP ${keep}`;
  }

  return `FROM ${source} | WHERE ${conditions.join(" AND ")} | SORT ${timestamp} DESC | LIMIT ${limit} | KEEP ${keep}`;
};

export const buildTrcQuery = (
  environment: EnvironmentConfig,
  logId: string,
  startTime: string,
  endTime: string
): string => {
  const source = index(environment.txntrcIndex);
  return `FROM ${source} | WHERE @timestamp >= "${literal(startTime)}" AND @timestamp < "${literal(endTime)}" AND ecp.log.id == "${literal(logId)}" | SORT ecp.log.timestamp ASC | LIMIT 20000 | KEEP ecp.log.timestamp, ecp.log.application, ecp.log.level, message`;
};

export const buildTraceQuery = (
  environment: EnvironmentConfig,
  traceId: string,
  startTime: string,
  endTime: string
): string => {
  const source = index(environment.apmIndex ?? "traces-apm*");
  return `FROM ${source} | WHERE trace.id == "${literal(traceId)}" AND @timestamp >= "${literal(startTime)}" AND @timestamp < "${literal(endTime)}" | SORT @timestamp ASC | LIMIT 20000 | KEEP @timestamp, trace.id, span.*, transaction.*, processor.event, service.*`;
};

export const pageRows = <T>(rows: T[], page: number, pageSize: number): T[] => {
  const start = Math.max(0, page - 1) * pageSize;
  return rows.slice(start, start + pageSize);
};

File: src-tauri/src/query_builders.rs
use crate::domain::{EnvironmentConfig, LogKind, SearchInput, SearchStatus, display_fields};

const MAX_PAGE_DEPTH: usize = 10_000;

fn literal(value: &str) -> String {
    value
        .trim()
        .replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', " ")
}

fn index(value: &str) -> Result<&str, String> {
    let allowed = value.chars().all(|character| {
        character.is_ascii_alphanumeric() || matches!(character, '.' | '_' | ',' | '*' | '-')
    });
    if !allowed || value.is_empty() {
        return Err("索引名称包含不允许的字符".to_string());
    }
    Ok(value)
}

fn time_range(field: &str, input: &SearchInput) -> String {
    format!(
        "{field} >= \"{}\" AND {field} < \"{}\"",
        literal(&input.start_time),
        literal(&input.end_time)
    )
}

fn add_like(conditions: &mut Vec<String>, field: &str, value: Option<&str>) {
    if let Some(value) = value.filter(|value| !value.trim().is_empty()) {
        conditions.push(format!("{field} LIKE \"*{}*\"", literal(value)));
    }
}

fn add_exact_or_like(
    conditions: &mut Vec<String>,
    field: &str,
    value: Option<&str>,
    looks_complete: fn(&str) -> bool,
) {
    let Some(value) = value.filter(|value| !value.trim().is_empty()) else {
        return;
    };
    let value = value.trim();
    if looks_complete(value) {
        conditions.push(format!("{field} == \"{}\"", literal(value)));
        return;
    }
    conditions.push(format!("{field} LIKE \"*{}*\"", literal(value)));
}

fn full_transaction_id(value: &str) -> bool {
    let normalized = value.to_ascii_lowercase();
    value.len() >= 24
        && [".p_", ".u_", ".s_", ".d_"]
            .iter()
            .any(|marker| normalized.contains(marker))
}

fn full_trace_id(value: &str) -> bool {
    (16..=64).contains(&value.len()) && value.chars().all(|character| character.is_ascii_hexdigit())
}

fn resolve_index<'a>(
    input: &'a SearchInput,
    environment: &'a EnvironmentConfig,
) -> Result<&'a str, String> {
    match input.kind {
        LogKind::Transaction => index(&environment.txnlst_index),
        LogKind::Application | LogKind::Ecp => index(&environment.applog_index),
        LogKind::Generic => {
            let selected = input
                .index
                .as_deref()
                .ok_or_else(|| "通用日志必须选择索引".to_string())?;
            let allowed = [
                environment.txnlst_index.as_str(),
                environment.txntrc_index.as_str(),
                environment.applog_index.as_str(),
                environment.apm_index.as_deref().unwrap_or("traces-apm*"),
            ];
            if !allowed.contains(&selected) {
                return Err("该索引未在当前环境中配置".to_string());
            }
            index(selected)
        }
    }
}

fn transaction_conditions(input: &SearchInput) -> Vec<String> {
    // UAT leaves ecp.txn.timestamp empty on part of the transaction stream.
    // The legacy Java client filters this log type by the canonical ingest time.
    let mut conditions = vec![time_range("@timestamp", input)];
    add_exact_or_like(
        &mut conditions,
        "ecp.txn.id",
        input.txn_id.as_deref(),
        full_transaction_id,
    );
    add_exact_or_like(
        &mut conditions,
        "ecp.txn.trace",
        input.trace_id.as_deref(),
        full_trace_id,
    );
    add_like(&mut conditions, "ecp.txn.no", input.txn_no.as_deref());
    add_like(
        &mut conditions,
        "ecp.txn.business",
        input.business.as_deref(),
    );
    add_like(&mut conditions, "ecp.txn.service", input.service.as_deref());
    add_like(
        &mut conditions,
        "ecp.txn.message.code",
        input.message_code.as_deref(),
    );
    add_like(
        &mut conditions,
        "ecp.txn.message.info",
        input.message_info.as_deref(),
    );
    add_like(&mut conditions, "ecp.txn.node", input.node.as_deref());

    match input.status {
        Some(SearchStatus::Success) => {
            conditions.push("RIGHT(ecp.txn.message.code, 5) == \"00000\"".to_string());
        }
        Some(SearchStatus::Fail) => {
            conditions.push("RIGHT(ecp.txn.message.code, 5) != \"00000\"".to_string());
        }
        _ => {}
    }
    if let Some(duration) = input.min_duration_ms.filter(|duration| *duration > 0) {
        conditions.push(format!("TO_INTEGER(ecp.txn.duration) >= {duration}"));
    }
    conditions
}

fn log_conditions(input: &SearchInput) -> Vec<String> {
    let timestamp = if input.kind == LogKind::Generic {
        "@timestamp"
    } else {
        "ecp.log.timestamp"
    };
    let mut conditions = vec![time_range(timestamp, input)];
    add_like(
        &mut conditions,
        "ecp.log.application",
        input.application.as_deref(),
    );
    add_like(&mut conditions, "ecp.log.level", input.level.as_deref());
    if input.kind == LogKind::Ecp {
        add_like(&mut conditions, "ecp.log.file", input.file.as_deref());
    }

    if let Some(keyword) = input
        .keyword
        .as_deref()
        .filter(|value| !value.trim().is_empty())
    {
        let fields: &[&str] = if input.kind == LogKind::Generic {
            &[
                "message",
                "ecp.log.application",
                "ecp.log.thread",
                "trace.id",
            ]
        } else {
            &["message", "ecp.log.thread", "trace.id", "host.name"]
        };
        let keyword = literal(keyword);
        let parts = fields
            .iter()
            .map(|field| format!("{field} LIKE \"*{keyword}*\""))
            .collect::<Vec<_>>();
        conditions.push(format!("({})", parts.join(" OR ")));
    }
    conditions
}

pub fn build_search_query(
    input: &SearchInput,
    environment: &EnvironmentConfig,
    export_all: bool,
) -> Result<String, String> {
    let source = resolve_index(input, environment)?;
    let conditions = if input.kind == LogKind::Transaction {
        transaction_conditions(input)
    } else {
        log_conditions(input)
    };
    let timestamp = match input.kind {
        LogKind::Generic => "@timestamp",
        _ => "ecp.log.timestamp",
    };
    let limit = if export_all {
        20_000
    } else {
        (input.page * input.page_size).min(MAX_PAGE_DEPTH)
    };
    let keep = display_fields(input.kind).join(", ");
    if input.kind == LogKind::Transaction {
        return Ok(format!(
            "FROM {source} | WHERE {} | SORT @timestamp DESC | LIMIT {limit} | EVAL ecp.txn.timestamp = COALESCE(ecp.txn.timestamp, @timestamp) | KEEP {keep}",
            conditions.join(" AND ")
        ));
    }

    Ok(format!(
        "FROM {source} | WHERE {} | SORT {timestamp} DESC | LIMIT {limit} | KEEP {keep}",
        conditions.join(" AND ")
    ))
}

pub fn build_trc_query(
    environment: &EnvironmentConfig,
    log_id: &str,
    start_time: &str,
    end_time: &str,
) -> Result<String, String> {
    let source = index(&environment.txntrc_index)?;
    Ok(format!(
        "FROM {source} | WHERE @timestamp >= \"{}\" AND @timestamp < \"{}\" AND ecp.log.id == \"{}\" | SORT ecp.log.timestamp ASC | LIMIT 20000 | KEEP ecp.log.timestamp, ecp.log.application, ecp.log.level, message",
        literal(start_time),
        literal(end_time),
        literal(log_id)
    ))
}

pub fn build_trace_query(
    environment: &EnvironmentConfig,
    trace_id: &str,
    start_time: &str,
    end_time: &str,
) -> Result<String, String> {
    let source = index(environment.apm_index.as_deref().unwrap_or("traces-apm*"))?;
    Ok(format!(
        "FROM {source} | WHERE trace.id == \"{}\" AND @timestamp >= \"{}\" AND @timestamp < \"{}\" | SORT @timestamp ASC | LIMIT 20000 | KEEP @timestamp, trace.id, span.*, transaction.*, processor.event, service.*",
        literal(trace_id),
        literal(start_time),
        literal(end_time)
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn environment() -> EnvironmentConfig {
        EnvironmentConfig {
            name: "test".into(),
            kibana_url: "https://kibana.example.test".into(),
            username: "reader".into(),
            password: "secret".into(),
            txnlst_index: "logs-ecp.txn.lst*".into(),
            txntrc_index: "logs-ecp.txn.trc*".into(),
            applog_index: "logs-ecp.log.*".into(),
            apm_index: None,
            allow_insecure_tls: None,
        }
    }

    fn input() -> SearchInput {
        SearchInput {
            environment: "test".into(),
            kind: LogKind::Transaction,
            start_time: "2026-09-01T00:00:00.000Z".into(),
            end_time: "2026-09-02T00:00:00.000Z".into(),
            page: 2,
            page_size: 50,
            index: None,
            txn_id: Some("approve-\"credit".into()),
            trace_id: None,
            txn_no: None,
            business: None,
            service: None,
            message_code: None,
            message_info: None,
            status: Some(SearchStatus::Fail),
            min_duration_ms: Some(120),
            node: None,
            keyword: None,
            level: None,
            file: None,
            application: None,
        }
    }

    #[test]
    fn preserves_transaction_rules() {
        let query = build_search_query(&input(), &environment(), false).unwrap();
        assert!(query.starts_with("FROM logs-ecp.txn.lst*"));
        assert!(query.contains("ecp.txn.id LIKE \"*approve-\\\"credit*\""));
        assert!(query.contains("RIGHT(ecp.txn.message.code, 5) != \"00000\""));
        assert!(query.contains("TO_INTEGER(ecp.txn.duration) >= 120"));
        assert!(query.contains("@timestamp >= \"2026-09-01T00:00:00.000Z\""));
        assert!(query.contains("EVAL ecp.txn.timestamp = COALESCE(ecp.txn.timestamp, @timestamp)"));
        assert!(query.contains("SORT @timestamp DESC | LIMIT 100 | EVAL"));
        assert!(query.contains("| KEEP ecp.txn.timestamp"));
    }

    #[test]
    fn uses_exact_matching_for_complete_identifiers() {
        let mut input = input();
        input.txn_id = Some("approve-trsFundTransferbui.u_0_7809041716470000000003".into());
        input.trace_id = Some("92d406f1f29331e8d0f273ef9faebc35".into());

        let query = build_search_query(&input, &environment(), false).unwrap();
        assert!(
            query.contains(
                "ecp.txn.id == \"approve-trsFundTransferbui.u_0_7809041716470000000003\""
            )
        );
        assert!(query.contains("ecp.txn.trace == \"92d406f1f29331e8d0f273ef9faebc35\""));
    }

    #[test]
    fn rejects_unconfigured_generic_index() {
        let mut input = input();
        input.kind = LogKind::Generic;
        input.index = Some("secrets-*".into());
        assert_eq!(
            build_search_query(&input, &environment(), false).unwrap_err(),
            "该索引未在当前环境中配置"
        );
    }

    #[test]
    fn limits_trace_payload_to_rendered_fields() {
        let query = build_trace_query(
            &environment(),
            "92d406f1f29331e8d0f273ef9faebc35",
            "2026-09-01T00:00:00.000Z",
            "2026-09-02T00:00:00.000Z",
        )
        .unwrap();
        assert!(query.contains(
            "LIMIT 20000 | KEEP @timestamp, trace.id, span.*, transaction.*, processor.event, service.*"
        ));
    }
}
